lib/semantics/resolve-names.cc

```diff
@@ -1038,11 +1038,267 @@ class ConstructVisitor : public DeclarationVisitor {
+  }
+
+  bool Pre(const parser::OpenMPThreadprivate &x) {
+    PushScope(Scope::Kind::Block, nullptr);
+    const auto &list{std::get<parser::OmpObjectList>(x.t)};
+    ResolveOmpObjectList(list, Symbol::Flag::OmpThreadprivate);
+    PopScope();
+    return false;
+  }
+
+  bool Pre(const parser::OmpClause::Shared &x) {
+    ResolveOmpObjectList(x.v, Symbol::Flag::OmpShared);
+    return false;
+  }
+  bool Pre(const parser::OmpClause::Private &x) {
+    ResolveOmpObjectList(x.v, Symbol::Flag::OmpPrivate);
+    return false;
+  }
+  bool Pre(const parser::OmpClause::Firstprivate &x) {
+    ResolveOmpObjectList(x.v, Symbol::Flag::OmpFirstPrivate);
+    return false;
+  }
+  bool Pre(const parser::OmpClause::Lastprivate &x) {
+    ResolveOmpObjectList(x.v, Symbol::Flag::OmpLastPrivate);
+    return false;
+  }
+
+  void add_dataSharingAttributeObject(const Symbol *object) {
+    dataSharingAttributeObjects_.insert(object);
+  }
+  void clear_dataSharingAttributeObject() {
+    dataSharingAttributeObjects_.clear();
+  }
+  bool find_dataSharingAttributeObject(const Symbol *);
+
+protected:
+  // TODO: resolve variables referenced in the OpenMP region
+  void ResolveOmpObjectList(const parser::OmpObjectList &, Symbol::Flag);
+  void ResolveOmpObject(const parser::OmpObject &, Symbol::Flag);
+  Symbol *ResolveOmp(const parser::Name &, Symbol::Flag);
+  Symbol *ResolveOmp(Symbol &, Symbol::Flag);
+  Symbol *ResolveOmpCommonBlockName(const parser::Name *);
+  Symbol *DeclarePrivateAccessEntity(const parser::Name &, Symbol::Flag);
+  Symbol *DeclarePrivateAccessEntity(Symbol &, Symbol::Flag);
+  Symbol *DeclareOrMarkOtherAccessEntity(const parser::Name &, Symbol::Flag);
+  Symbol *DeclareOrMarkOtherAccessEntity(Symbol &, Symbol::Flag);
+  void CheckMultipleAppearances(
+      const parser::Name &, const Symbol *, Symbol::Flag);
+
+private:
+  std::set<const Symbol *> dataSharingAttributeObjects_;  // on one directive
+};
+
+bool OmpVisitor::find_dataSharingAttributeObject(const Symbol *object) {
+  auto it{dataSharingAttributeObjects_.find(object)};
+  if (it != dataSharingAttributeObjects_.end()) {
+    return true;
+  }
+  return false;
+}
+
+Symbol *OmpVisitor::ResolveOmpCommonBlockName(const parser::Name *name) {
+  if (auto *prev{name ? currScope().parent().FindCommonBlock(name->source)
+                      : nullptr}) {
+    name->symbol = prev;
+    return prev;
+  } else {
+    return nullptr;
+  }
+}
+
+void OmpVisitor::ResolveOmpObjectList(
+    const parser::OmpObjectList &ompObjectList, Symbol::Flag ompFlag) {
+  for (const auto &ompObject : ompObjectList.v) {
+    ResolveOmpObject(ompObject, ompFlag);
+  }
+}
+
+void OmpVisitor::ResolveOmpObject(
+    const parser::OmpObject &ompObject, Symbol::Flag ompFlag) {
+  const auto &kind{std::get<parser::OmpObject::Kind>(ompObject.t)};
+  const auto &designator{std::get<parser::Designator>(ompObject.t)};
+  const auto *name{GetDesignatorNameIf(designator)};
```

Comment:
This would be less confusing if the parse tree matched the grammar better. Can `OmpObject` have `std::variant<Designator, Name>`. The first alternative is a normal Fortran designator, the second is a common block name. Then you don't have to worry about what to do if `kind` says common block but `designator` is not a `Name`.

Reply:
That is very true, I will see what I can do.

Reply:
This was done in PR#753 and adjustment was made for this PR.